src/components/time-range-dropdown.tsx

```diff
@@ -1,62 +1,126 @@
+import {
+  CUSTOM_TIME_RANGE_KEY,
+  formatTimeRange,
+  timeOptionKeyFromRange,
+  timeRangeFromDuration,
+  timeRangeOptions,
+} from '../time-range';
+import { TimeRangeSelectModal } from './time-range-select-modal';
 
-const DEFAULT_TIME_RANGE = '1h';
+const DEFAULT_DURATION_KEY = '1h';
 const STORED_TIME_RANGE_KEY = 'logging-view-plugin.time-range';
 
 interface TimeRangeDropdownProps {
-  initialValue?: string;
-  onChange?: (offset: number) => void;
+  value?: TimeRange;
+  onChange?: (timeRange: TimeRange) => void;
   isDisabled?: boolean;
 }
 
+const getSelectedOptionIndex = ({
+  storedTimeRange,
+  timeRangeValue,
+}: {
+  storedTimeRange?: TimeRange | null;
+  timeRangeValue?: TimeRange;
+}): number => {
+  let durationKey: string | undefined;
+
+  if (timeRangeValue) {
+    durationKey = timeOptionKeyFromRange(timeRangeValue);
+  }
+
+  if (storedTimeRange) {
+    durationKey = timeOptionKeyFromRange(storedTimeRange);
+  }
+
+  if (!durationKey) {
+    durationKey = DEFAULT_DURATION_KEY;
+  }
+
+  return timeRangeOptions.findIndex((option) => option.key === durationKey) ?? 1;
+};
+
 export const TimeRangeDropdown: React.FC<TimeRangeDropdownProps> = ({
   onChange,
-  initialValue = DEFAULT_TIME_RANGE,
+  value,
   isDisabled = false,
 }) => {
-  const [storedTimeRange, setStoredTimeRange] = useLocalStorage(STORED_TIME_RANGE_KEY);
+  const [isTimeRangeModalOpen, setIsTimeRangeModalOpen] = React.useState(false);
+  const [storedTimeRange, setStoredTimeRange] = useLocalStorage<TimeRange>(STORED_TIME_RANGE_KEY);
+
   const [isOpen, setIsOpen] = React.useState(false);
   const [selectedIndex, setSelectedIndex] = React.useState<number>(
-    [
-      timeRangeOptions.findIndex((option) => option.key === storedTimeRange),
-      timeRangeOptions.findIndex((option) => option.key === initialValue),
-      1,
-    ].filter((index) => index >= 0)[0],
+    getSelectedOptionIndex({ storedTimeRange, timeRangeValue: value }),
   );
 
+  React.useEffect(() => {
+    setSelectedIndex(getSelectedOptionIndex({ timeRangeValue: value }));
+  }, [value]);
+
   const handleSelectedValue = (index: number) => () => {
     setIsOpen(false);
-    setSelectedIndex(index);
 
-    const span = timeRangeOptions[index].span;
-    onChange?.(span);
-    setStoredTimeRange(timeRangeOptions[index].key);
+    const { key } = timeRangeOptions[index];
+
+    if (key === CUSTOM_TIME_RANGE_KEY) {
+      setIsTimeRangeModalOpen(true);
+    } else {
+      setSelectedIndex(index);
+      const timeRange = timeRangeFromDuration(timeRangeOptions[index].key);
+      onChange?.(timeRange);
+      setStoredTimeRange(timeRange);
+    }
+  };
+
+  const handleCustomRangeSelected = (customRange: TimeRange) => {
+    setSelectedIndex(0);
+    setIsTimeRangeModalOpen(false);
+    onChange?.(customRange);
+    setStoredTimeRange(customRange);
   };
 
   const toggleIsOpen = () => {
     setIsOpen(!isOpen);
   };
 
+  const handleCloseModal = () => {
```

Comment:
I think it's good to use a simple `useBoolean` hook to avoid needing to define simple handlers like this. (See https://github.com/openshift/console/blob/master/frontend/public/components/monitoring/hooks/useBoolean.ts)

Reply:
Agreed, I'll do it in a separate PR as some parts unrelated to this PR might change.